Use embeddable v2

stacey-gammon opened the pull request:
# For reviewers

`dashboard_state_manager` used to be the primary way the redux store was synced with app state.  Redux was essentially replaced with the dashboard embeddable container. However rather than put this in dashboard state manager, the logic is split up between dashboard state manager and dashboard_app_controller.  It's pretty illogical atm, but given that this entire flow needs some serious clean up, and that app state is going away, as is angular, I simply took the path of least resistance.

Moving forward I could see some next steps being:
 - Adding a migration from `SavedDashboardPanel` to `DashboardPanelState`.  The former is what is stored in the `PanelsJSON`, the latter is what the `DashboardEmbeddableContainer` expects.  Any previous ad hoc migration logic that used to exist inside the dashboard grid has been moved into the dashboard app (so the new  `DashboardEmbeddableContainer` expects data in a certain format, users of the system are required to make sure they send it over in that particular format). 

# Dev Docs

Any plugins written on the old embeddable infrastructure will need to be updated to use the new embeddable infrastructure.  The Embeddables API allows you to add your own custom UI components to a dashboard, without having to go through the visualization plugin system.  This infrastructure is highly unstable at the moment.  Expect this API to change frequently over the ensuing months.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2436/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2543/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2547/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2580/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2601/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2901/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2910/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3007/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3193/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3311/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3321/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3490/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3621/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3633/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3771/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3780/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3799/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3816/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3834/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3876/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3999/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4033/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4060/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4076/)

stacey-gammon commented:
I think I'm close with tests.  Everything is passing but one which looks like flaky ci:
```
13:19:34 Running as SYSTEM
13:19:35 ERROR: Issue with creating launcher for agent kibana-ci-immutable-centos-1561567245720335843. The agent has not been fully initialized yet
13:19:35 [EnvInject] - Loading node environment variables.
13:19:35 [EnvInject] - Preparing an environment for the build.
13:19:35 [EnvInject] - Keeping Jenkins system variables.
13:19:35 [EnvInject] - Keeping Jenkins build variables.
13:19:35 [EnvInject] - Node root path is not available. Likely node is offline. Skipping injection
13:19:35 Building remotely on kibana-ci-immutable-centos-1561567245720335843 (immutable virtual centos-7 x86_64 swarm linux centos docker)FATAL: Cannot retrieve the node's workspace for hudson.matrix.MatrixProject@3eae0d80[elastic+kibana+pull-request]. Most probably the node is offline
13:19:35 java.io.IOException: Cannot retrieve the node's workspace for hudson.matrix.MatrixProject@3eae0d80[elastic+kibana+pull-request]. Most probably the node is offline
13:19:35 	at hudson.matrix.MatrixRun$MatrixRunExecution.getParentWorkspaceLease(MatrixRun.java:173)
13:19:35 	at hudson.matrix.MatrixRun$MatrixRunExecution.decideWorkspace(MatrixRun.java:187)
13:19:35 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:488)
13:19:35 	at hudson.model.Run.execute(Run.java:1818)
13:19:35 	at hudson.matrix.MatrixRun.run(MatrixRun.java:153)
13:19:35 	at hudson.model.ResourceController.execute(ResourceController.java:97)
13:19:35 	at hudson.model.Executor.run(Executor.java:429)
13:19:35 Sending e-mails to: infra-root+build@elastic.co
13:19:36 ERROR: Build step failed with exception
13:19:36 java.lang.NullPointerException
13:19:36 	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:187)
13:19:36 	at com.google.jenkins.plugins.storage.AbstractUpload.perform(AbstractUpload.java:160)
13:19:36 	at com.google.jenkins.plugins.storage.AbstractUpload.perform(AbstractUpload.java:133)
13:19:36 	at com.google.jenkins.plugins.storage.GoogleCloudStorageUploader.perform(GoogleCloudStorageUploader.java:110)
13:19:36 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
13:19:36 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:741)
13:19:36 	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
13:19:36 	at hudson.model.Build$BuildExecution.post2(Build.java:186)
13:19:36 	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
13:19:36 	at hudson.model.Run.execute(Run.java:1843)
13:19:36 	at hudson.matrix.MatrixRun.run(MatrixRun.java:153)
13:19:36 	at hudson.model.ResourceController.execute(ResourceController.java:97)
13:19:36 	at hudson.model.Executor.run(Executor.java:429)
13:19:36 Build step 'Google Cloud Storage Plugin' marked build as failure
13:19:36 ERROR: Step ‘Publish JUnit test result report’ failed: no workspace for elastic+kibana+pull-request/JOB=x-pack-ciGroup1,node=immutable #4076
13:19:36 Finished: FAILURE
```

jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4096/)

elasticmachine commented:
Pinging @elastic/kibana-app-arch

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4756/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4766/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4801/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4826/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4871/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6065/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6090/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6102/)

stacey-gammon commented on src/legacy/core_plugins/embeddable_api/public/context_menu_actions/build_eui_context_menu_panels.ts:
```diff
@@ -104,12 +104,10 @@ function convertPanelActionToContextMenuItem({
-      closeMenu();
-    };
-  }
+  menuPanelItem.onClick = () => {
```
Change was necessary or the link actions that navigated the user to another page would result in the pop up menu still showing.

cchaos commented on src/legacy/core_plugins/kibana/public/discover/embeddable/_index.scss:
```diff
@@ -0,0 +1,11 @@
+   * 2. Force a better looking scrollbar
+   */
+.embPanel {
+   .kbnDocTable__container {
```
This style block isn't being applied. You'll need to move these styles into a `_embeddable.scss` file which would get imported here, then import this index file in https://github.com/elastic/kibana/blob/9cf5e2c9f65dc5f60a76cf8c03cea5e6377e687f/src/legacy/core_plugins/kibana/public/discover/_index.scss

stacey-gammon replied:
Good catch!

streamich commented on src/legacy/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable_factory.tsx:
```diff
@@ -0,0 +1,191 @@
+      );
+    } catch (e) {
+      console.error(e); // eslint-disable-line no-console
+      return new ErrorEmbeddable(e, input, parent);
```
Why returning instead of throwing?

stacey-gammon replied:
I don't want an error to break an entire dashboard so it's better to return an error embeddable that shows the user the issue.

I suppose I could also have factories throw errors and make the container handle catching and converting into an error embeddable. Though it's kind of nice that the container can assume it will get an embeddable of some sort.

One common error here is if there is no saved object for that id.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6395/)

stacey-gammon commented on src/legacy/core_plugins/kibana/public/discover/embeddable/search_embeddable_factory.ts:
```diff
@@ -46,35 +53,64 @@ export class SearchEmbeddableFactory extends EmbeddableFactory {
-  ) {
-    const editUrl = this.getEditPath(id);
-    const editable = capabilities.get().discover.save as boolean;
+  public async createFromSavedObject(
```
we use await earlier... I'm going to actually just  convert this to using await. it's possible whatever required this before doesn't anymore because of all the changes and there are no errors and no UI issues (that I notice) when I change it over.

stacey-gammon commented on src/legacy/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable.ts:
```diff
@@ -159,17 +191,16 @@ export class VisualizeEmbeddable extends Embeddable {
-    this.query = containerState.query;
-    this.filters = containerState.filters;
+  public render(domNode: HTMLElement) {
+    this.timeRange = _.cloneDeep(this.input.timeRange);
```
`this.timeRange` is used above to check if the previous timeRange passed in has changed.  Could possibly be a more first class way of knowing what exactly changed when a new input is passed in...

and yea, deeply cloned because somewhere something was mutating them and then messing up the comparison of "did something change, should I update my observers?".  Kind of like redux tree, input should never be mutated directly.  I couldn't figure out where though!

streamich commented on src/legacy/core_plugins/kibana/public/dashboard/lib/embeddable_saved_object_converters.test.ts:
```diff
@@ -0,0 +1,134 @@
+    type: 'search',
+  };
+
+  expect(convertPanelStateToSavedDashboardPanel(dashboardPanel)).toEqual({
```
Yes, `hasOwnProperty` much nicer.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6448/)

stacey-gammon commented:
@nreese - any chance you can just do some UI testing to make sure I didn't break anything with the maps embeddable with this PR?

nreese commented:
I am getting the following error.

I did a clean install `kbn clean && kbn bootstrap`
I installed all three sample data sets
I tried to open dashboard application.

Nothing rendered and the console had the following errors.

<img width="800" alt="Screen Shot 2019-07-10 at 8 26 20 AM" src="https://user-images.githubusercontent.com/373691/60977315-775bd780-a2ec-11e9-86c1-e4d251d276c4.png">

stacey-gammon commented:
I can't seem to repro that @nreese.  Can you share the URL? I tried copying the url for the sample flights dashboard on master and navigating directly to that on my branch and it worked fine.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6666/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6694/)

stacey-gammon commented:
Nathan found a bug which while digging deeper exposed a couple issues.

There was a logic bug where `if( dirty)` in `dashboard_app_controller.tsx` was always returning true because dirty was actually an object, not a boolean. Not sure why typescript didn't catch this but I guess it's technically okay to do that (thought it would have some "always true" warning).

Fixing this actually exposed another issue because filters aren't changing appState inside `dashboardStateManager.applyFilters` which should be what cause app state to change, which should then cause a `refreshDashboardContainer` call.  The reason this worked in the old code is because there was a `this._pushFiltersToStore();` to handle the case of pinned filters not causing an update to redux because app state didn't change.

adding `refreshDashboardContainer();` in `dashboard_app_controller` where the queryFilter is being watched should fix this:

```
    const updateSubscription = queryFilter.getUpdates$().subscribe({
      next: () => {
        $scope.model.filters = queryFilter.getFilters();
        dashboardStateManager.applyFilters($scope.model.query, $scope.model.filters);
        ***refreshDashboardContainer();
      },
    });
``` 

Now this still doesn't actually fix the original error.  Now if you try to load a dashboard with a filter on it, this happens:
1. The dashboard container gets initialized with no filters because `queryFilter.getFilters()` is empty (though appState has filters in it... I'm guessing some subscription hasn't been fired yet to fill it).

2. Dashboard container tells visualize factory to grab the saved object to construct a new VisualizeEmbeddable.

3. Control goes back to the dashboard app, and `updateQueryAndFetch` fires, still `queryFilter.getFilters()` is empty so we are telling it to fetch the wrong data (and since we don't have query cancellation, this is not great - I suspect this is happening on master too).

4. `updateQueryAndFetch` changes dashboard container input (via `lastReloadRequest`).  At least in one step through `dashboardContainer` wasn't finished constructing so it's null and the re-fetch doesn't get fired.

5. `queryFilter.getUpdates$()` is fired. Now we tell dashboard container to update with filters.

6. Control goes back to the embeddable factory which is trying to create the embeddable for the container.  It was given `initialInput` from the container, it creates a new embeddable using it. In the Embeddable constructor there is a listener that listens to its parent input changes.  It gets fired immediately (just the way the SubjectBehavior listener works), it grabs the newest changes from its parent, and updates its input, before returning. 

6. The original `getInputForChild` that was grabbed earlier now doesn't match the embeddables input and it looks like the embeddable factory creation tried to override some container information, and that data gets stuffed on explicitInput when it shouldn't. 

I added a new jest test to catch this situation.

So... why is this logic in the second half of `Container.onPanelAdded` even in there?  I think it was for an earlier iteration with exploring the embeddable factories that request user data when creating their embeddable.  Initially this logic was inside the `create` function but this was later moved externally into a specific `EmbeddableFactory.getExplicitInput` function which I think works much better than this confusing, and fragile comparison to determine explicit input.  So, thankfully, I should just be able to drop all that unnecessary code and the logic is easier to follow and the bug goes away.

I also added some extra code in the embeddable explorer to make sure the one sample embeddable I have that's using this logic that an embeddable creation process that requests explicit input works correctly and it does.

<img width="1418" alt="Screen Shot 2019-07-10 at 4 24 26 PM" src="https://user-images.githubusercontent.com/16563603/61002434-8a43cb80-a32f-11e9-88d4-c0a30c27f70a.png">

Whew.

Stay tuned for changes.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6745/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6772/)

stacey-gammon commented:
retest

stacey-gammon commented:
Well, at least I proved that somehow my test changes are causing failures, completely unrelated to this PR - see https://github.com/elastic/kibana/pull/40826.  Fails with

```
            └-> "before all" hook
                 │ERROR browser[SEVERE] http://localhost:5620/bundles/commons.bundle.js 0:820568 Error: Uncaught IndexSizeError: Failed to execute 'getImageData' on 'CanvasRenderingContext2D': The source width is 0. (http://localhost:5620/built_assets/dlls/vendors.bundle.dll.js:34)
                 │          at window.onerror (http://localhost:5620/bundles/commons.bundle.js:1:952189)
                 │ERROR browser[SEVERE] http://localhost:5620/built_assets/dlls/vendors.bundle.dll.js 33:10592 Uncaught DOMException: Failed to execute 'getImageData' on 'CanvasRenderingContext2D': The source width is 0.
                 │ERROR browser[SEVERE] http://localhost:5620/bundles/commons.bundle.js 0:952188 Uncaught Error: Uncaught IndexSizeError: Failed to execute 'getIma…:5620/built_assets/dlls/vendors.bundle.dll.js:34)
                 └- ✖ fail: "dashboard app using current data dashboard filtering adding a filter that excludes all data "before all" hook"
                 │      Error: expected testSubject(globalLoadingIndicator-hidden) to exist
                 │       at TestSubjects.existOrFail (test/functional/services/test_subjects.ts:56:15)
                 │       at process._tickCallback (internal/process/next_tick.js:68:7)
                 │ 
```

:( but :) I didn't kill it... I guess

stacey-gammon commented:
I don't want to block this PR on that issue, so I have a separate PR out that would add a selenium test that would catch it but is causing other errors on ci on master - https://github.com/elastic/kibana/pull/40826.

I did add a jest test anyway in this PR, but functionally testing will have to be manual. 

 - save a dashboard
 - add a filter to that dashboard (don't save it)
- go to dashboard listing page
- re-open that dashboard
- apply filter and see if it works.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6784/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6788/)

stacey-gammon commented:
grrr.  Unfortunately adding:

```
    // update root source when filters update
    const updateSubscription = queryFilter.getUpdates$().subscribe({
      next: () => {
        $scope.model.filters = queryFilter.getFilters();
        dashboardStateManager.applyFilters($scope.model.query, $scope.model.filters);
        refreshDashboardContainer();
      },
    });
```

caused another issue during "create new visualization" flow (luckily caught by selenium tests). 

Partially there was a bug in the embeddable API (another jest test added), in the situation that a panel was added and then removed before the embeddable finished loading, the container would attempt to access that panel and an error would be thrown.

I fixed that, but there was still an issue in the dashboard app code because of all the async updating coming from multiple state changes.  What happened, assuming you just went through the "add new visualization flow" and hit "save and add to dashboard" is:

- Dashboard loads, when it's finished, it checks the URL for the special parameter indicating to add a new embeddable to itself.  This process starts but it's async.
- before the embeddable finishes loading, the async filter update is hit and the dashboard container is updated, this time with the filter, *but also with the panel missing*.  This can be fixed by not calling the generic `refreshDashboardContainer()` which updates the container with all differences detected, but instead, just calls `dashboardContainer.updateInput({ filters: $scope.model.filters });` directly.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6797/)

nreese commented on x-pack/legacy/plugins/maps/public/embeddable/map_embeddable.js:
```diff
@@ -176,37 +171,38 @@ export class MapEmbeddable extends Embeddable {
-    if (!_.isEqual(this._embeddableConfig.openTOCDetails, openTOCDetails)) {
-      embeddableConfigChanged = true;
-      this._embeddableConfig.openTOCDetails = openTOCDetails;
+    if (!this.input.isLayerTOCOpen
```
`!this.input.isLayerTOCOpen` is not needed in this check and will fire updateInput every time input. isLayerTOCOpen is false.

Should just be
`if (this.input.isLayerTOCOpen !== isLayerTOCOpen) `

stacey-gammon commented:
Failed on `Code Explore Repository Explore a repository Click file/directory on the file tree`.  I just merged master so I'm thinking this is flaky, and looks like it's failing in other prs too.

<img width="806" alt="Screen Shot 2019-07-12 at 9 36 14 AM" src="https://user-images.githubusercontent.com/16563603/61132043-8f685e00-a488-11e9-9e1f-bc0cc71f755c.png">

jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7108/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7114/)

stacey-gammon commented:
I think the dashboard filtering before all hook is flaky.

```
-: dashboard filtering
[00:06:47]               └-> "before all" hook
[00:06:47]               └-> "before all" hook
[00:06:47]                 │ debg gotoDashboardLandingPage
[00:06:47]                 │ debg onDashboardLandingPage
[00:06:47]                 │ debg TestSubjects.exists(dashboardLandingPage)
[00:06:47]                 │ debg Find.existsByDisplayedByCssSelector('[data-test-subj~="dashboardLandingPage"]') with timeout=5000
[00:06:47]               └-: adding a filter that excludes all data
[00:06:47]                 └-> "before all" hook
[00:06:47]                 └-> "before all" hook
[00:06:47]                   │ debg TestSubjects.exists(newItemButton)
[00:06:47]                   │ debg Find.existsByDisplayedByCssSelector('[data-test-subj~="newItemButton"]') with timeout=2500
[00:06:50]                   │ debg TestSubjects.click(createDashboardPromptButton)
[00:06:50]                   │ debg Find.clickByCssSelector('[data-test-subj~="createDashboardPromptButton"]') with timeout=10000
[00:06:50]                   │ debg Find.findByCssSelector('[data-test-subj~="createDashboardPromptButton"]') with timeout=10000
[00:07:01]                   │ debg --- retry.try error: Waiting for element to be located By(css selector, [data-test-subj~="createDashboardPromptButton"])
[00:07:01]                   │      Wait timed out after 10190ms
[00:07:01]                   │ debg Find.findByCssSelector('[data-test-subj~="createDashboardPromptButton"]') with timeout=10000
[00:07:11]                   │ debg --- retry.try error: Waiting for element to be located By(css selector, [data-test-subj~="createDashboardPromptButton"])
[00:07:11]                   │      Wait timed out after 10020ms
[00:07:12]                   │ debg Find.findByCssSelector('[data-test-subj~="createDashboardPromptButton"]') with timeout=10000
[00:07:22]                   │ debg --- retry.try error: Waiting for element to be located By(css selector, [data-test-subj~="createDashboardPromptButton"])
[00:07:22]                   │      Wait timed out after 10529ms
[00:07:23]                   │ debg Find.findByCssSelector('[data-test-subj~="createDashboardPromptButton"]') with timeout=10000
[00:07:33]                   │ debg --- retry.try error: Waiting for element to be located By(css selector, [data-test-subj~="createDashboardPromptButton"])
[00:07:33]                   │      Wait timed out after 10400ms
[00:07:34]                   │ debg Find.findByCssSelector('[data-test-subj~="createDashboardPromptButton"]') with timeout=10000
[00:07:44]                   │ debg --- retry.try error: Waiting for element to be located By(css selector, [data-test-subj~="createDashboardPromptButton"])
```

<img width="1257" alt="Screen Shot 2019-07-12 at 10 59 33 AM" src="https://user-images.githubusercontent.com/16563603/61137996-9eeda400-a494-11e9-90a4-4c36c5987606.png">

not sure if this PR exacerbates the issue, but given the screenshot and output above I don't see why it would. I don't know why it's failing either, the screenshot looks like the `createDashboardPromptButton` should have been found.

stacey-gammon commented:
Actually I think I see why that test is flaky. 

```
    async clickNewDashboard() {
      // newItemButton button is only visible when dashboard listing table is displayed
      // (at least one dashboard).
      const exists = await testSubjects.exists('newItemButton');
      if (exists) {
        return await testSubjects.click('newItemButton');
      }

      // If no dashboards exist, then newItemButton to create a new dashboard.
      return await this.clickCreateDashboardPrompt();
    }
```

the page probably didn't finish loading so `newItemButton` didn't exist, which meant it tried to click the create new dashboard from the prompt, not the button in the table.  This can be fixed in another PR, it's unrelated.

stacey-gammon commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7127/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7141/)
